### DISCUSSION TITLE
What's the benefit of using mutation?

### DISCUSSION THREAD

**Discussion by buglessir** [TanStack/query#3158] [General] [answered]
I could find many helpful use cases for using the `useQuery` in app, but I can't find any for `useMutation`
For example, we have a form to add a row in database (with POST method), that API returns 2xx or 5xx codes without any `response.text`
So, what shall I do for these kinds of API calls? there is no data for caching... should I even use `mutation` for this use case?
Actually, I'm confuse about exact use cases for the `mutation`...
Should I use that only for changing cached data? or should use it for all of the POST, PUT etc. requests?
I mean, something like changing `Products List` after adding new product by mutation and using `queryClient.invalidateQueries` to invalidate old data??

**Comment by TkDodo** [comment, marked as answer]
There are a couple of things that mutations give you, like:

- tracking of error and loading state so that you can show a loading spinner / disable a button etc. while the mutation is going on
- callbacks like `onSuccess` and `onError` so that you can `invalidate` your list query after you've added an entry so that it actually show up in the list
- an easy way to perform [optimistic updates](https://react-query.tanstack.com/guides/optimistic-updates) so that you can add the entry to your list query even before the mutation has finished.

For more info, please read https://tkdodo.eu/blog/mastering-mutations-in-react-query
